libcxx/test/std/ranges/range.adaptors/range.stride.view/types.h

```diff
@@ -0,0 +1,259 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef TEST_STD_RANGES_RANGE_ADAPTORS_RANGE_STRIDE_TYPES_H
+#define TEST_STD_RANGES_RANGE_ADAPTORS_RANGE_STRIDE_TYPES_H
+
+#include <iterator>
+#include <ranges>
+#include <type_traits>
+
+#include "__iterator/concepts.h"
+#include "__ranges/common_view.h"
+#include "test_iterators.h"
+#include "test_range.h"
+
+// Concepts
+
+template <typename Iter>
+concept IterDifferable = requires(Iter& t) { t - t; };
+
+// Iterators
+
+template <class Derived, std::input_iterator Iter = int*, bool IsSized = false>
+  requires((!IsSized) || (IsSized && IterDifferable<Iter>))
+struct InputIterBase {
+  using iterator_concept  = std::input_iterator_tag;
+  using iterator_category = std::input_iterator_tag;
+  using value_type        = typename std::iterator_traits<Iter>::value_type;
+  using difference_type   = typename std::iterator_traits<Iter>::difference_type;
+
+  Iter value_{};
+
+  constexpr InputIterBase()                                      = default;
+  constexpr InputIterBase(const InputIterBase&)                  = default;
+  constexpr InputIterBase(InputIterBase&&)                       = default;
+  constexpr InputIterBase& operator=(const InputIterBase& other) = default;
+  constexpr InputIterBase& operator=(InputIterBase&& other)      = default;
+  constexpr explicit InputIterBase(Iter value) : value_(value) {}
+
+  constexpr value_type operator*() const { return *value_; }
+  constexpr Derived& operator++() {
+    value_++;
+    return static_cast<Derived&>(*this);
+  }
+  constexpr Derived operator++(int) {
+    auto nv = *this;
+    value_++;
+    return nv;
+  }
+  friend constexpr bool operator==(const Derived& left, const Derived& right) { return left.value_ == right.value_; }
+  friend constexpr difference_type operator-(const Derived& left, const Derived& right)
+    requires IsSized
+  {
+    return left.value_ - right.value_;
+  }
+};
+
+struct UnsizedInputIterator : InputIterBase<UnsizedInputIterator> {};
+static_assert(std::input_iterator<UnsizedInputIterator>);
+static_assert(!std::sized_sentinel_for<UnsizedInputIterator, UnsizedInputIterator>);
+
+struct SizedInputIterator : InputIterBase<SizedInputIterator, int*, true> {
+  using InputIterBase::InputIterBase;
+};
+static_assert(std::input_iterator<SizedInputIterator>);
+static_assert(std::sized_sentinel_for<SizedInputIterator, SizedInputIterator>);
+
+// Don't move/hold the iterator itself, copy/hold the base
+// of that iterator and reconstruct the iterator on demand.
+// May result in aliasing (if, e.g., Iterator is an iterator
```

Comment:
This type only seems to be used once -- if that's the case, I'd suggest moving it to the file that uses it so that its purpose is more obvious.